File: games/pong/src/main/kotlin/com/neoutils/engine/games/pong/Ball.kt
package com.neoutils.engine.games.pong

import com.neoutils.engine.math.Rect
import com.neoutils.engine.math.Vec2
import com.neoutils.engine.physics.BoxCollider
import com.neoutils.engine.physics.Collider
import com.neoutils.engine.render.Color
import com.neoutils.engine.render.Renderer
import com.neoutils.engine.scene.Node2D
import kotlin.math.sign
import kotlin.random.Random

class Ball(
    val size: Float = 16f,
    val initialSpeed: Float = 280f,
    val maxSpeed: Float = 560f,
    val speedupPerHit: Float = 1.05f,
    var fieldCenter: Vec2,
    private val random: Random = Random.Default,
    private val onScore: (Goal.Side) -> Unit = {},
) : Node2D() {

    var velocity: Vec2 = randomVelocity(initialSpeed, random)
        private set

    val collider: BoxCollider = object : BoxCollider(Vec2(size, size)) {
        override fun onCollide(other: Collider) {
            handleCollision(other)
        }
    }

    init {
        addChild(collider)
        reset(serveToward = if (random.nextBoolean()) 1f else -1f)
    }

    override fun onUpdate(dt: Float) {
        transform = transform.copy(position = transform.position + velocity * dt)
    }

    override fun onRender(renderer: Renderer) {
        val center = transform.position + Vec2(size / 2f, size / 2f)
        renderer.drawCircle(center, radius = size / 2f, color = Color.WHITE, filled = true)
    }

    private fun handleCollision(other: Collider) {
        when {
            other is Goal -> {
                val scorer = if (other.side == Goal.Side.Left) Goal.Side.Right else Goal.Side.Left
                onScore(scorer)
                reset(serveToward = if (other.side == Goal.Side.Left) 1f else -1f)
            }
            other is Wall -> velocity = velocity.copy(y = -velocity.y)
            other is PaddleCollider -> {
                val paddleBounds = other.bounds()
                val paddleCenterY = paddleBounds.top + paddleBounds.size.y / 2f
                val ballCenterY = transform.position.y + size / 2f
                val relative = ((ballCenterY - paddleCenterY) /
                    (paddleBounds.size.y / 2f)).coerceIn(-1f, 1f)

                val newSpeed = (velocity.length * speedupPerHit).coerceAtMost(maxSpeed)
                val horizontalSign = if (velocity.x > 0f) -1f else 1f
                val maxAngleRad = (kotlin.math.PI / 3f).toFloat() // ±60°
                val angle = relative * maxAngleRad
                velocity = Vec2(
                    horizontalSign * newSpeed * kotlin.math.cos(angle),
                    newSpeed * kotlin.math.sin(angle),
                )

                // Nudge ball out of paddle to prevent re-collision next tick.
                val ballPos = transform.position
                val ballRight = ballPos.x + size
                val ballLeft = ballPos.x
                val shift = if (horizontalSign < 0f) paddleBounds.left - ballRight - 0.5f
                else paddleBounds.right - ballLeft + 0.5f
                transform = transform.copy(position = ballPos.copy(x = ballPos.x + shift))
            }
        }
    }

    fun reset(serveToward: Float) {
        transform = transform.copy(position = fieldCenter - Vec2(size / 2f, size / 2f))
        // Wider angle range so the AI paddle has visible vertical work to do
        // and the ball doesn't degenerate into a horizontal line.
        val angle = (random.nextFloat() - 0.5f) * 1.4f // ~ ±0.7 rad (~40deg)
        val sx = if (serveToward >= 0f) 1f else -1f
        velocity = Vec2(
            sx * initialSpeed * kotlin.math.cos(angle),
            initialSpeed * kotlin.math.sin(angle),
        )
    }

    private fun clampSpeed(v: Vec2, max: Float): Vec2 {
        val l = v.length
        if (l <= max) return v
        val s = max / l
        return Vec2(v.x * s, v.y * s)
    }
}

private fun randomVelocity(speed: Float, random: Random): Vec2 {
    val angle = (random.nextFloat() - 0.5f) * 0.6f
    val sx = if (random.nextBoolean()) 1f else -1f
    return Vec2(sx * speed * kotlin.math.cos(angle), speed * kotlin.math.sin(angle))
        .let { Vec2(sign(it.x).let { s -> if (s == 0f) 1f else s } * kotlin.math.abs(it.x), it.y) }
}

File: games/pong/src/main/resources/pong/scripts/ball.py
# extends BoxCollider

import math
import random as _random

ballSize: float = 16.0
initialSpeed: float = 280.0
maxSpeed: float = 560.0
speedupPerHit: float = 1.05
fieldCenter: Vec2 = Vec2(400.0, 300.0)

# Emitted with the side string ("Left" or "Right") when the ball touches a
# goal collider. The PongScene wires this signal up to the scoreboards.
scored: Signal = signal(str)


def _ready(self):
    self._velocity = Vec2(0.0, 0.0)
    self._scored_this_tick = False
    self.size = Vec2(self.ballSize, self.ballSize)
    if not getattr(self, '_initialized', False):
        _reset(self, 1.0 if _random.random() > 0.5 else -1.0)
        self._initialized = True


def _physics_process(self, dt):
    self._scored_this_tick = False
    self.size = Vec2(self.ballSize, self.ballSize)
    pos = self.position
    self.position = Vec2(pos.x + self._velocity.x * dt, pos.y + self._velocity.y * dt)


def _draw(self, renderer):
    wp = self.world().position
    center = Vec2(wp.x + self.ballSize / 2.0, wp.y + self.ballSize / 2.0)
    renderer.drawCircle(center, self.ballSize / 2.0, Color(1.0, 1.0, 1.0, 1.0), True, 1.0)


def _on_collide(self, other):
    if self._scored_this_tick:
        return
    name = other.name
    if name == "leftGoal":
        self.scored.emit("Right")
        _reset(self, 1.0)
        self._scored_this_tick = True
        return
    if name == "rightGoal":
        self.scored.emit("Left")
        _reset(self, -1.0)
        self._scored_this_tick = True
        return
    # Paddle hit: the collider lives one level under a Paddle (named "left" /
    # "right" in scene.json). Bouncing geometry uses the paddle bounds so the
    # angle reflects where on the paddle the ball struck.
    parent = other.parent
    if parent is not None and parent.name in ("left", "right"):
        paddle_bounds = other.bounds()
        paddle_center_y = paddle_bounds.top + paddle_bounds.size.y / 2.0
        ball_center_y = self.position.y + self.ballSize / 2.0
        rel = (ball_center_y - paddle_center_y) / (paddle_bounds.size.y / 2.0)
        if rel < -1.0:
            rel = -1.0
        elif rel > 1.0:
            rel = 1.0
        vlen = self._velocity.length
        new_speed = vlen * self.speedupPerHit
        if new_speed > self.maxSpeed:
            new_speed = self.maxSpeed
        h_sign = -1.0 if self._velocity.x > 0.0 else 1.0
        max_angle = math.pi / 3.0
        angle = rel * max_angle
        self._velocity = Vec2(h_sign * new_speed * math.cos(angle), new_speed * math.sin(angle))
        ball_pos = self.position
        ball_right = ball_pos.x + self.ballSize
        ball_left = ball_pos.x
        if h_sign < 0.0:
            shift = paddle_bounds.left - ball_right - 0.5
        else:
            shift = paddle_bounds.right - ball_left + 0.5
        self.position = Vec2(ball_pos.x + shift, ball_pos.y)
        return
    # Fall-through: walls (topWall / bottomWall) flip the vertical component.
    self._velocity = Vec2(self._velocity.x, -self._velocity.y)


def reset(self, serve_toward):
    _reset(self, serve_toward)


def _reset(self, serve_toward):
    self.size = Vec2(self.ballSize, self.ballSize)
    self.position = Vec2(
        self.fieldCenter.x - self.ballSize / 2.0,
        self.fieldCenter.y - self.ballSize / 2.0,
    )
    angle = (_random.random() - 0.5) * 1.4
    sx = 1.0 if serve_toward >= 0.0 else -1.0
    self._velocity = Vec2(
        sx * self.initialSpeed * math.cos(angle),
        self.initialSpeed * math.sin(angle),
    )
